Applications/Spire/Source/Ui/TextBox.cpp

```diff
@@ -162,101 +97,106 @@ connection TextBox::connect_submit_signal(
   return m_submit_signal.connect(slot);
 }
 
-void TextBox::changeEvent(QEvent* event) {
-  switch(event->type()) {
-    case QEvent::ReadOnlyChange:
-    case QEvent::FontChange:
+bool TextBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_line_edit) {
+    switch(event->type()) {
     case QEvent::StyleChange:
-      elide_text();
+      if(m_old_style_sheet != styleSheet()) {
+        m_old_style_sheet = styleSheet();
+        update_display_text();
+      }
+      break;
+    case QEvent::ReadOnlyChange:
+    case QEvent::EnabledChange:
+      update_display_text();
+      break;
+    case QEvent::FocusIn:
+      if(!is_read_only()) {
+        auto reason = static_cast<QFocusEvent*>(event)->reason();
+        if(reason != Qt::ActiveWindowFocusReason &&
+            reason != Qt::PopupFocusReason) {
+          update_display_text();
+        }
+      }
+      break;
+    case QEvent::FocusOut:
+      {
+        auto reason = static_cast<QFocusEvent*>(event)->reason();
+        if(reason != Qt::ActiveWindowFocusReason &&
+            reason != Qt::PopupFocusReason) {
+          update_display_text();
+        }
+      }
+      break;
+    case QEvent::KeyPress:
+      if(!is_read_only() && static_cast<QKeyEvent*>(event)->key() ==
+          Qt::Key_Escape) {
+        m_current = m_submission;
+        m_line_edit->setText(m_current);
+        m_current_signal(m_current);
+      }
       break;
-  }
-  QLineEdit::changeEvent(event);
-}
-
-void TextBox::focusInEvent(QFocusEvent* event) {
-  if(!isReadOnly()) {
-    auto reason = event->reason();
-    if(reason != Qt::ActiveWindowFocusReason &&
-      reason != Qt::PopupFocusReason) {
-      QLineEdit::setText(m_text);
     }
   }
-  QLineEdit::focusInEvent(event);
-}
-
-void TextBox::focusOutEvent(QFocusEvent* event) {
-  auto reason = event->reason();
-  if(reason != Qt::ActiveWindowFocusReason && reason != Qt::PopupFocusReason) {
-    elide_text();
-  }
-  QLineEdit::focusOutEvent(event);
-}
-
-void TextBox::keyPressEvent(QKeyEvent* event) {
-  if(event->key() == Qt::Key_Escape) {
-    m_text = m_submitted_text;
-    QLineEdit::setText(m_text);
-    m_current_signal(m_text);
-  }
-  QLineEdit::keyPressEvent(event);
+  return QWidget::eventFilter(watched, event);
 }
 
 void TextBox::resizeEvent(QResizeEvent* event) {
-  elide_text();
-  QLineEdit::resizeEvent(event);
+  update_display_text();
+  QWidget::resizeEvent(event);
 }
 
 QSize TextBox::sizeHint() const {
   return scale(160, 30);
 }
 
 void TextBox::on_editing_finished() {
-  if(!isReadOnly()) {
-    m_submitted_text = m_text;
-    m_submit_signal(m_submitted_text);
+  if(!is_read_only()) {
+    m_submission = m_current;
+    m_submit_signal(m_submission);
   }
 }
 
 void TextBox::on_selection_changed() {
-  if(isReadOnly()) {
-    auto text = selectedText();
+  if(is_read_only()) {
+    auto text = m_line_edit->selectedText();
     if(text.endsWith(QChar(0x2026)) || text.endsWith(QStringLiteral("..."))) {
```

Comment:
What happens if the actual current text ends with a "..."? The user can type in something like "Hello...".

Reply:
Great catch. When the text is “Hello…” and the textbox is read-only, it will trigger the selection change event infinitely. It has no problem in non-read-only mode. I am going to fix the bug.